lib/cartopy/mpl/scale_bar.py

```diff
@@ -0,0 +1,411 @@
+            - us-yd U.S. Surveyor's Yard
+            - us-ch U.S. Surveyor's Chain
+            - us-mi U.S. Surveyor's Statute Mile
+            - ind-yd Indian Yard
+            - ind-ft Indian Foot
+            - ind-ch Indian Chain
+        zorder: Integer, matplotlib plotting order (higher values are plotted
+            later.)
+        bar_length: Float value (in 'units' units) to override the default bar
+            length as described above.  If this parameter is set, then
+            divisions must ALSO be set.  Caveat emptor.
+        divisions: Integer value to override the default number of bar
+            divisions as described above.  Allowable divisions values
+            are: [2,3,4,5,6]. If this parameter is set, then bar_length must
+            ALSO be set.  Caveat emptor.
+        """
+
+        hasbar = bar_length is not None
+        hasdiv = divisions is not None
+        if hasbar + hasdiv != 0 and hasbar + hasdiv != 2:
+            raise Exception('You must specify BOTH bar_length AND divisions.')
+
+        if divisions is not None:
+            if divisions not in [2, 3, 4, 5, 6]:
+                raise Exception(
+                    'Divisions must be an integer in the range 2-6.')
+
+        # check units
+        if units not in UNITS:
+            raise Exception(
+                'Input scale bar units must be either "km" or "mi"')
+
+        # check position
+        if pos not in ['ll', 'lr', 'ul', 'ur']:
+            raise Exception("Input scale bar position must be one of "
+                            "'ll', 'lr', 'ul', 'ur'")
+
+        # Verify that the input axes are in fact GeoAxes
+        if not isinstance(ax, GeoAxes):
+            raise Exception('Input axes object must be a Cartopy GeoAxes.')
+
+        # Get the left, right, bottom, top edges in data units
+        self._xmin, self._xmax = ax.get_xlim()
+        self._ymin, self._ymax = ax.get_ylim()
+
+        # stuff input parameters into class attributes
+        self._position = pos
+        self._pad_x = padx
+        self._pad_y = pady
+        self._units = units
+        self._zorder = zorder
+        self._font_size = font_size
+
+        # Get the factor that converts the data units into meters
+        if not isinstance(ax.projection, cartopy.crs.PlateCarree):
+            if 'units' in ax.projection.proj4_params:
+                proj_units = ax.projection.proj4_params['units']
+                self._dataunits_to_meters = UNITS[proj_units]
+            else:
+                proj_units = 'm'
+                self._dataunits_to_meters = UNITS[proj_units]
+        else:
+            # this is PlateCaree, which has no projection
+            proj_units = 'dd'
+            clat = self._ymin + (self._ymax - self._ymin) / 2
+            self._dataunits_to_meters = DD2METERS * np.cos(np.radians(clat))
+            # true at center of map
+        self._meters_to_dataunits = 1 / self._dataunits_to_meters
```

Comment:
This whole block is unnecessary (I think). We can make use of cartopy's geodesic calculations.